Allow loading of llvm plugins on nightly

ZuseZ4 opened the pull request:
Based on a discussion in  #82734 / with @wsmoses.
Mainly moves [this](https://github.com/wsmoses/rust/commit/0149bc4e7e596005c665b132877abebe5258a0f6) behind a -Z flag, so it can only be used on nightly,
as requested by @nagisa in https://github.com/rust-lang/rust/issues/82734#issuecomment-835863940

This change allows loading of llvm plugins like Enzyme.
Right now it also requires a shared library LLVM build of rustc for symbol resolution.

```rust
// test.rs
extern { fn __enzyme_autodiff(_: usize, ...) -> f64; }

fn square(x : f64) -> f64 {
   return x * x;
}

fn main() {
   unsafe {
      println!("Hello, world {} {}!", square(3.0), __enzyme_autodiff(square as usize, 3.0));
   }
}
```
```
./rustc test.rs -Z llvm-plugins="./LLVMEnzyme-12.so" -C passes="enzyme"
./test
Hello, world 9 6!
```

I will try to figure out how to simplify the usage and get this into stable in a later iteration, 
but having this on nightly will already help testing further steps.

ZuseZ4 committed:
allow loading of llvm plugins on nightly

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @varkor (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

varkor commented:
r? @nagisa

nagisa commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -129,6 +129,16 @@ unsafe fn configure_llvm(sess: &Session) {
+        if res.is_null() {
+            println!("{}", CStr::from_ptr(libc::dlerror()).to_string_lossy().into_owned());
+        }
+        println!("{:p}", res);
```
No debug `println`s in the compiler codebase, please. Consider `debug!`.

ZuseZ4 replied:
Done

nagisa commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -129,6 +129,16 @@ unsafe fn configure_llvm(sess: &Session) {
 
+    for plugin in &sess.opts.debugging_opts.llvm_plugins {
+        let path = CString::new(plugin.as_bytes()).unwrap();
+        let res = libc::dlopen(path.as_ptr(), libc::RTLD_LAZY | libc::RTLD_GLOBAL);
```
Consider using a [`DynamicLibrary`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/dynamic_lib/struct.DynamicLibrary.html) instead.

ZuseZ4 replied:
Done

nagisa commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -129,6 +129,16 @@ unsafe fn configure_llvm(sess: &Session) {
+        let path = CString::new(plugin.as_bytes()).unwrap();
+        let res = libc::dlopen(path.as_ptr(), libc::RTLD_LAZY | libc::RTLD_GLOBAL);
+        if res.is_null() {
+            println!("{}", CStr::from_ptr(libc::dlerror()).to_string_lossy().into_owned());
```
This should probably fail the compilation session.

ZuseZ4 replied:
Replaced by bug!, as used in other locations in this file.

ZuseZ4 committed:
addressing feedback

ZuseZ4 committed:
Add missing use

ZuseZ4 commented:
I think I pushed the second commit from the wrong build dir, but now it should be all right, unless you have some further change requests?
I will look on how we can expose the missing symbols by default so we don't need the shared lib llvm build anymore, but this will be another PR.

nagisa commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -129,6 +132,15 @@ unsafe fn configure_llvm(sess: &Session) {
 
+    for plugin in &sess.opts.debugging_opts.llvm_plugins {
+        let path = Path::new(plugin);
+        let res = DynamicLibrary::open(path);
```
I think you want to `mem::forget(res)` here, otherwise the library will get almost immediately closed in `Drop` hook of `DynamicLibrary`.

ZuseZ4 replied:
Adding mem::forget makes sense since we want the plugin to be available during the entire compilation process.
However, for my testcase it has been also working before. Is this expected (probably because we have a small test program and the compilation finished before the dropping takes effect (?)), or is this an issue?

thomcc replied:
Often dlclose does nothing, but you absolutely should not rely on it doing nothing.

nagisa commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -129,6 +132,15 @@ unsafe fn configure_llvm(sess: &Session) {
+        let path = Path::new(plugin);
+        let res = DynamicLibrary::open(path);
+        match res {
+            Ok(_) => debug!("configure_llvm: {}", plugin),
```
nit: tracing already includes information about functions and modules that the debug call is in. This should perhaps say more about what's being done here precisely:

```suggestion
            Ok(_) => debug!("LLVM plugin loaded {} ({})", path.display(), plugin),
```

Or something along those lines.

ZuseZ4 replied:
Makes sense, I will update it in this case. But how about this section, is it outdated? https://rustc-dev-guide.rust-lang.org/compiler-debugging.html#logging-etiquette-and-conventions

`A loosely followed convention is to use debug!("foo(...)") at the start of a function foo and debug!("foo: ...") within the function.`

nagisa replied:
Huh, I wonder if that convention predates the use of tracing for logging. @oli-obk do you know?

ZuseZ4 committed:
Remove dropping of loaded plugins and better debug info

nagisa commented:
@bors r+ Nice!

bors commented:
:pushpin: Commit abdd24a0404e2ce7c879297da4aa66cf4e64ab79 has been approved by `nagisa`

<!-- @bors r=nagisa abdd24a0404e2ce7c879297da4aa66cf4e64ab79 -->
<!-- homu: {"type":"Approved","sha":"abdd24a0404e2ce7c879297da4aa66cf4e64ab79","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit abdd24a0404e2ce7c879297da4aa66cf4e64ab79 with merge a1411de9de38e0fed728874580218338160eb185...
<!-- homu: {"type":"BuildStarted","head_sha":"abdd24a0404e2ce7c879297da4aa66cf4e64ab79","merge_sha":"a1411de9de38e0fed728874580218338160eb185"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2922618378)
Approved by: nagisa
Pushing a1411de9de38e0fed728874580218338160eb185 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2922618378"},"merge_sha":"a1411de9de38e0fed728874580218338160eb185"} -->

bgat commented:
I'm trying to load and use my own LLVM plugin from rustc, and getting the following error, even though the very same .so file works when use it via opt:

**error: failed to run LLVM passes: Failed to load pass plugin/home/ubuntu/coast/projects/build/DWC/libDWC.so**

I'm using the legacy plugin format. Is the above related to the, "Right now it also requires a shared library LLVM build of rustc for symbol resolution" requirement?

ZuseZ4 commented:
what do you get from `opt --version` ?

I don't think that the latest LLVM (which rustc is building or downloading by default) does support legacy plugins at all.
I was actually about to propose to remove this legacy fallback one of these days.
Is there a reason why you can't move to the newPM?

ZuseZ4 commented:
I think in this case you should probably just take an older rust commit to build. I don't think that you will get a legacy commit working eith an up do date LLVM

CXWorks commented:
Hi @ZuseZ4 , I'm trying to use my own afl-pass with asan like below:

`RUSTFLAGS=" -Z llvm-plugins=./afl-llvm-pass.so -C passes=hello"  cargo  build --bin xxx`

But it seems doesn't work: `error: failed to run LLVM passes: unknown pass name 'hello'`

ZuseZ4 commented:
@CXWorks You should look at newer PRs mentioning LLVM plugins - My original plugin infra iirc got removed and replaced by a newer, since LLVM plugins for the new PassManager work a bit different. You would need to look at those PRs to figure out how to use them. For my own project I moved away from the plugin interface, since it was too limited, so I didn't follow closely.

CXWorks commented:
Thanks for your quick response!
